feat: Add pre and post workflow hook status

Fabianoshz opened the pull request:
Currently when a pre or post workflow hook is running there's no indication to the user. When running a lengthy task like using [terragrunt-atlantis-config](https://github.com/transcend-io/terragrunt-atlantis-config) on a big repository a PR may stay a long time without any return.

This tries to help users (especially those with low experience) to identity the current status of the operation.

Fabianoshz commented:
Still need to work on tests for this :sweat_smile:

nitrocode commented:
Nice job @Fabianoshz ! I'd love tests and I'd love to see the stdout output via the details link like we can for the terraform plan/applies.

Fabianoshz commented:
Hi @nitrocode, I'll start writing tests for this, about the stdout output maybe on a separated PR? I'm still familiarizing myself with Atlantis code base.

nitrocode commented:
A separate pr would be okay i suppose but in the mean time it would break consistency with what we currently have for the atlantis plan and apply workflows which do expose a details link. If you go the separate pr route, please do not forget to ass that functionality or we may have to revert the pr.

Thoughts? @runatlantis/maintainers

jamengual commented:
maybe keep it in the same PR since they will be related?

nitrocode commented:
It would be nice too if the status of the pre and post workflows is modifiable like @albertorm95 is attempting in pr https://github.com/runatlantis/atlantis/pull/2699. that pr will parae the output because the output is standard but in the case of pre and post workflows, there can be any script run, so it would be nice if we could set it to be a custom regex in the yaml to parse the output to bubble it up as the status. But of course this is just a bonus.

Fabianoshz commented:
I'll look how to add stdout to the status link in this PR, I've messed around with the code base enough to fell confident about trying this. @nitrocode  I will try to add a way of setting the status too, I'm unsure about the regex though, what do you think about passing a environment variable to the command, expecting the user to push the output it wants to the variable and then we expose it later? Regex might break in ways user might not expect, also it would be harder for them to test things.

Fabianoshz commented:
@nitrocode Didn't managed to use a environment variable, instead I've passed an env with a filepath to the command so it can send the status to it. I just need to check the string size limit for this before sending it to the VCS client.

Fabianoshz commented:
Ok, managed to make the output work, it will add the link to the check:
![image](https://user-images.githubusercontent.com/14815404/207223728-f5113be8-8e8c-4a3a-9a4e-fdbc1b86e6f7.png)

And will output the stdout to the UI:
![image](https://user-images.githubusercontent.com/14815404/207223874-599d6332-e696-4ffd-8049-255680b9f27c.png)

Had to change a lot of things to make this work, and I have heavily coupled the hooks code to the functions that the jobs use to show the output too, I believe more changes will be required to make this more readable in the future, I think I'm gonna need more time to finish this.

Also the tests are broken (had to fix a bazillion times because I've kept changing things).

Fabianoshz commented:
Custom descriptions seems to work fine too:
![image](https://user-images.githubusercontent.com/14815404/207224989-0a4118de-73e7-4f18-acd2-705af625d149.png)

nitrocode commented on server/core/runtime/pre_workflow_hook_runner.go:
```diff
@@ -47,8 +51,20 @@ func (wh DefaultPreWorkflowHookRunner) Run(ctx models.WorkflowHookCommandContext
+
 	ctx.Log.Info("successfully ran %q in %q", command, path)
-	return string(out), nil
+	return string(out), strings.Trim(string(cmdOut), "\n"), nil
```
What's the difference between out and cmdOut?

nitrocode replied:
Thank you for the response.

Would you mind adding that in as a comment for the new `out` input var? At the moment, the naming convention of both are confusing to me. It's not clear from reading the code without having to dive in further.

Fabianoshz replied:
Sure, will do!

nitrocode commented on go.mod:
```diff
@@ -79,6 +79,8 @@ require (
 	cloud.google.com/go/storage v1.28.0 // indirect
 	github.com/Masterminds/goutils v1.1.1 // indirect
 	github.com/Masterminds/semver/v3 v3.2.0 // indirect
+	github.com/alecthomas/template v0.0.0-20190718012654-fb15b899a751 // indirect
```
This dependency hasnt been updated in 3 years

https://github.com/alecthomas/template

Can we manage without it? Or is it possible to use another more up to date dependency?

nitrocode commented on go.mod:
```diff
@@ -79,6 +79,8 @@ require (
 	github.com/Masterminds/goutils v1.1.1 // indirect
 	github.com/Masterminds/semver/v3 v3.2.0 // indirect
+	github.com/alecthomas/template v0.0.0-20190718012654-fb15b899a751 // indirect
+	github.com/alecthomas/units v0.0.0-20190924025748-f65c72e2690d // indirect
```
This dependency hasnt been updated in 1 year

https://github.com/alecthomas/units

Can we manage without it? Or is it possible to use another more up to date dependency?

nitrocode commented on server/events/post_workflow_hooks_command_runner.go:
```diff
@@ -89,13 +96,34 @@ func (w *DefaultPostWorkflowHooksCommandRunner) runHooks(
+			return err
+		}
+
+		if err := w.CommitStatusUpdater.UpdatePostWorkflowHook(ctx.Pull.BaseRepo, ctx.Pull, models.SuccessCommitStatus, hookDescription, runtimeDesc, url); err != nil {
```
Could we move this if block above line 117 and then remove lines 118 to 120?

Fabianoshz replied:
Line 114 marks the VCS check as pending, line 121 VCS marks the check as failed and line 127 marks VCS check as success. I don't think we can remove these lines as it would remove the VCS check update functionality for the workflow hooks.

![image](https://user-images.githubusercontent.com/14815404/207635387-3cf2f1ab-79a3-4471-84ae-ff5f61de0d1d.png)

Fabianoshz commented:
Hey @nitrocode, removed those dependencies, I didn't have need for them, not sure why I added it to be fair, I believe I'm still kinda far to finishing this, there's still way too much coupling with the jobs package which I think is not a good idea, so I might leave some more useless stuff in the code while I test things, but I will make sure the code has only exactly what it need by the end.

nitrocode commented:
@Fabianoshz without having to refactor, can this still be done in multiple PRs? Such as a PR to get the feature in and another PR to refactor to simplify it?

Fabianoshz commented:
@nitrocode sure is possible, I still need to fix some stuff to make this work correctly, but once is done I'm going to open a new PR to refactor the decoupling.

Fabianoshz commented:
@nitrocode @jamengual

I was having a hard time running the `preWorkflowHooksSetup` and `postWorkflowHooksSetup` tests because the workflow hook model is created inside these functions, and since I've added the HookID as a UUID I could not match this with the model created in tests with the model created in the functions because the UUID is unique. 

I've ended up adding this file: https://github.com/runatlantis/atlantis/blob/6c17dda94a2cfc02e240590bd30c6edd3cc69506/server/events/uuid_generator.go (which can be mocked to return static values) but I'm not so sure that it's fits the in `events` package.

Usually in other projects I would expect to have a repository that could be mocked with static values but there's no such thing in Atlantis and I didn't want to add a new pattern just because of this.

Fabianoshz commented:
Ok, I'm far more comfortable with this code now, still need some refactoring to decouple from the stuff on the `jobs` package though.

nitrocode commented on server/core/runtime/post_workflow_hook_runner.go:
```diff
@@ -47,8 +54,25 @@ func (wh DefaultPostWorkflowHookRunner) Run(ctx models.WorkflowHookCommandContex
+	// Read the value from the "outputFilePath" file
+	// to be returned as a custom description.
+	var customStatusOut []byte
+	if _, err := os.Stat(outputFilePath); err == nil {
```
If there is an err, then we should handle that, right? At the very least maybe log something to debug

GenPage replied:
We can let it silently fail. It basically means that the user didn't have a custom description to set as part of the workflow since the file doesn't exist.

nitrocode replied:
Why does the description have to be saved as an output file? Shouldn't we pass the description to the function instead of outputting a file on the disk and then reading it later?

Fabianoshz replied:
I'm using a file because right now the stdout is used for the details link. You gave the idea of a regex [here](https://github.com/runatlantis/atlantis/pull/2441#issuecomment-1345742992) but I believe that this would make testing kinda hard for users, so that's why I've added the file.

https://github.com/runatlantis/atlantis/blob/e176af2a0c5bff36a70390f183f710cde45bed15/runatlantis.io/docs/post-workflow-hooks.md?plain=1#L79

Fabianoshz replied:
Do you believe there's a better way of handling custom status? I've tried setting an environment variable but that didn't work, the [Cmd struct](https://pkg.go.dev/os/exec#Cmd) has no way of returning an environment variable, you can only set but not get from the process.

Fabianoshz replied:
Hmm, I just noticed that I can get the pid from the Cmd, that means that I probably can try to read `/proc/$pid/environ`, maybe that can enable the usage of environment variables?

nitrocode replied:
Let's do a file for now since it works.

nitrocode commented on server/core/runtime/pre_workflow_hook_runner.go:
```diff
@@ -47,8 +54,25 @@ func (wh DefaultPreWorkflowHookRunner) Run(ctx models.WorkflowHookCommandContext
+		return "", "", err
+	}
+
+	// Read the value from the "outputFilePath" file
```
This is a similar block to the post workflow function. Could we create a function that does the same logic and use it in both places?

GenPage replied:
Not unless we created a generic workflow_hook_runner interface to define functions for pre/post to implement, or we refactor them into one class.

nitrocode replied:
I meant a single function in a `utils.go` type of package to use the same block on line 60 in this file and the `post` workflow hook runner. I did not mean to update the interface for helper functions as that would be a much more involved refactor.

Fabianoshz replied:
We could add this to the `utils.go` but as you noticed post and pre workflow hooks are basically the same with different names, I believe a major refactor would be the better way to go and I'm willing to do this, if you wish I can start this on this PR or I can open another PR as you [suggested here](https://github.com/runatlantis/atlantis/pull/2441#issuecomment-1351694854).

nitrocode replied:
Ok that's fair enough. Let's refactor in the next pr

nitrocode commented on server/events/post_workflow_hooks_command_runner.go:
```diff
@@ -89,13 +102,33 @@ func (w *DefaultPostWorkflowHooksCommandRunner) runHooks(
 
-	for _, hook := range postWorkflowHooks {
-		_, err := w.PostWorkflowHookRunner.Run(ctx, hook.RunCommand, repoDir)
+	for i, hook := range postWorkflowHooks {
```
After reviewing this code, it seems like the post and pre workflow hooks are very similar. Perhaps in the future we can combine them so we do not need to copy pasta between post and pre golang files. This is probably what you were getting at with your follow up refactor pr.

Fabianoshz replied:
Yup, you got it right, I would also like to decouple the code from the `jobs` package, I had to couple it together to add the details link functionality.

nitrocode replied:
Ok let's refractor in a follow up pr

nitrocode reviewed (approved):
Looks pretty good to me.

cc @runatlantis/maintainers for more eyes

GenPage commented on server/core/runtime/post_workflow_hook_runner.go:
```diff
@@ -47,8 +54,25 @@ func (wh DefaultPostWorkflowHookRunner) Run(ctx models.WorkflowHookCommandContex
 	}
+
+	wh.OutputHandler.SendWorkflowHook(ctx, string(out), false)
+	wh.OutputHandler.SendWorkflowHook(ctx, "", true)
```
nitpick: can we format `string(out)` with a newline to avoid another function call?

GenPage commented on server/core/runtime/pre_workflow_hook_runner.go:
```diff
@@ -47,8 +54,25 @@ func (wh DefaultPreWorkflowHookRunner) Run(ctx models.WorkflowHookCommandContext
 	}
+
+	wh.OutputHandler.SendWorkflowHook(ctx, string(out), false)
+	wh.OutputHandler.SendWorkflowHook(ctx, "", true)
```
nitpick: same comment about new line and multiple function calls as above

GenPage reviewed (approved):
Great PR, just some nitpicks.

krrrr38 commented on runatlantis.io/docs/post-workflow-hooks.md:
```diff
@@ -74,4 +76,5 @@ command](custom-workflows.html#custom-run-command).
   * `USER_NAME` - Username of the VCS user running command, ex. `acme-user`. During an autoplan, the user will be the Atlantis API user, ex. `atlantis`.
   * `COMMENT_ARGS` - Any additional flags passed in the comment on the pull request. Flags are separated by commas and
     every character is escaped, ex. `atlantis plan -- arg1 arg2` will result in `COMMENT_ARGS=\a\r\g\1,\a\r\g\2`.
+  * `OUTPUT_FILE` - An output file you can use to customize the success or failure status. ex. `echo 'failure' > $OUTPUT_FILE`.
```
Is it better to use `OUTPUT_STATUS_FILE`? Cause `OUTPUT_FILE` means too large and atlantis may out put more contents in the future like https://github.com/runatlantis/atlantis/issues/936.

krrrr38 reviewed (commented):
nice work.

jamengual reviewed (commented):
@Fabianoshz when you have some time, resolve the discussions/feedback 
otherwise it looks good to me

nitrocode commented on runatlantis.io/docs/post-workflow-hooks.md:
```diff
@@ -74,4 +76,5 @@ command](custom-workflows.html#custom-run-command).
   * `USER_NAME` - Username of the VCS user running command, ex. `acme-user`. During an autoplan, the user will be the Atlantis API user, ex. `atlantis`.
   * `COMMENT_ARGS` - Any additional flags passed in the comment on the pull request. Flags are separated by commas and
     every character is escaped, ex. `atlantis plan -- arg1 arg2` will result in `COMMENT_ARGS=\a\r\g\1,\a\r\g\2`.
+  * `OUTPUT_STATUS_FILE` - An output file you can use to customize the success or failure status. ex. `echo 'failure' > $OUTPUT_STATUS_FILE`.
```
```suggestion
  * `OUTPUT_STATUS_FILE` - An output file to customize the success or failure status. ex. `echo 'failure' > $OUTPUT_STATUS_FILE`.
```

nitrocode commented on runatlantis.io/docs/post-workflow-hooks.md:
```diff
@@ -74,4 +76,5 @@ command](custom-workflows.html#custom-run-command).
   * `USER_NAME` - Username of the VCS user running command, ex. `acme-user`. During an autoplan, the user will be the Atlantis API user, ex. `atlantis`.
   * `COMMENT_ARGS` - Any additional flags passed in the comment on the pull request. Flags are separated by commas and
     every character is escaped, ex. `atlantis plan -- arg1 arg2` will result in `COMMENT_ARGS=\a\r\g\1,\a\r\g\2`.
+  * `OUTPUT_STATUS_FILE` - An output file you can use to customize the success or failure status. ex. `echo 'failure' > $OUTPUT_STATUS_FILE`.
```
Is this specific to pre and post workflows? Or is it for any status? I ask because the description doesn't specify.

Fabianoshz replied:
It is specific to the pre and post workflows, this is the post-workflow-hooks doc, so I believe this should well scoped. I can add this description if you wish though.

nitrocode commented on runatlantis.io/docs/pre-workflow-hooks.md:
```diff
@@ -58,5 +60,6 @@ command](custom-workflows.html#custom-run-command).
   * `USER_NAME` - Username of the VCS user running command, ex. `acme-user`. During an autoplan, the user will be the Atlantis API user, ex. `atlantis`.
   * `COMMENT_ARGS` - Any additional flags passed in the comment on the pull request. Flags are separated by commas and
       every character is escaped, ex. `atlantis plan -- arg1 arg2` will result in `COMMENT_ARGS=\a\r\g\1,\a\r\g\2`.
+  * `OUTPUT_STATUS_FILE` - An output file you can use to customize the success or failure status. ex. `echo 'failure' > $OUTPUT_STATUS_FILE`.
```
```suggestion
  * `OUTPUT_STATUS_FILE` - An output file to customize the success or failure status. ex. `echo 'failure' > $OUTPUT_STATUS_FILE`.
```

nitrocode commented:
@runatlantis/maintainers ready for a second review :pray:

krrrr38 commented on server/core/runtime/pre_workflow_hook_runner.go:
```diff
@@ -47,8 +54,24 @@ func (wh DefaultPreWorkflowHookRunner) Run(ctx models.WorkflowHookCommandContext
+		}
 	}
+
+	wh.OutputHandler.SendWorkflowHook(ctx, fmt.Sprintf("%s\n", string(out)), true)
```
ah, in this case JobID means HookID. so no problem. sorry.

krrrr38 reviewed (approved):
lgtm

nitrocode commented:
@Fabianoshz nice job! This will be very helpful in the next release.

Thanks for the additional set of eyes @krrrr38 and @GenPage !

ribejara-te commented:
Hey, this is a great feature. Thanks!
